vpr/src/route/router_lookahead_map_utils.cpp

```diff
@@ -0,0 +1,619 @@
+#include "router_lookahead_map_utils.h"
+
+#include "globals.h"
+#include "vpr_context.h"
+#include "vtr_math.h"
+#include "vtr_time.h"
+#include "route_common.h"
+#include "route_timing.h"
+
+static void dijkstra_flood_to_wires(int itile, RRNodeId inode, util::t_src_opin_delays& src_opin_delays);
+static void dijkstra_flood_to_ipins(RRNodeId node, util::t_chan_ipins_delays& chan_ipins_delays);
+
+static vtr::Point<int> pick_sample_tile(t_physical_tile_type_ptr tile_type, vtr::Point<int> start);
+
+// Constants needed to reduce the bounding box when expanding CHAN wires to reach the IPINs
+#define X_OFFSET 2
+#define Y_OFFSET 2
+
+// Maximum dijkstra expansions when exploring the CHAN --> IPIN connections
+#define MAX_EXPANSION_LEVEL 1
+
+#define DIRECT_CONNECT_SPECIAL_SEG_TYPE -1;
+
+namespace util {
+
+PQ_Entry::PQ_Entry(
+    int set_rr_node_ind,
+    int switch_ind,
+    float parent_delay,
+    float parent_R_upstream,
+    float parent_congestion_upstream,
+    bool starting_node,
+    float Tsw_adjust) {
+    this->rr_node_ind = set_rr_node_ind;
+
+    auto& device_ctx = g_vpr_ctx.device();
+    this->delay = parent_delay;
+    this->congestion_upstream = parent_congestion_upstream;
+    this->R_upstream = parent_R_upstream;
+    if (!starting_node) {
+        float Tsw = device_ctx.rr_switch_inf[switch_ind].Tdel;
+        Tsw += Tsw_adjust;
+        VTR_ASSERT(Tsw >= 0.f);
+        float Rsw = device_ctx.rr_switch_inf[switch_ind].R;
+        float Cnode = device_ctx.rr_nodes[set_rr_node_ind].C();
+        float Rnode = device_ctx.rr_nodes[set_rr_node_ind].R();
+
+        float T_linear = 0.f;
+        if (device_ctx.rr_switch_inf[switch_ind].buffered()) {
+            T_linear = Tsw + Rsw * Cnode + 0.5 * Rnode * Cnode;
+        } else { /* Pass transistor */
+            T_linear = Tsw + 0.5 * Rsw * Cnode;
```

Comment:
Nag, nag ... good to check this.

Reply:
Where is the delay function defined?

Reply:
https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/98810f87d33737ffecea841d177ae6142aa288c9/vpr/src/route/route_tree_timing.cpp#L564

That basically computes delay assuming the total capacitance loading a node (Cdownstream) is already computed (the computation of that is elsewhere in the same file).  The common case is that all switches are buffered, and hence Cdownstream is just the node capacitance plus the Cinternal capacitance of switches that are turned on and have that same node as their input.

I guess the lookahead is really modeling fanout = 1 connections, so no other switches besides the one downstream one that goes to the next node will be on. So adding the Cinternal of that downstream switch * (Rwire/2 + Rswitch_driving_this_node) is probably all that should be added.  If Cinternal is 0 it doesn't matter, but to model architectures that choose to model this loading effect (by have Rwire, Rswitch and Cinternal be non-zero) it would be good to include it, even though it probably isn't a very big effect.